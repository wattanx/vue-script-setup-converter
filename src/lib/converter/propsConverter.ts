import {
  CallExpression,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
  PropertyAssignment,
  SyntaxKind,
  Node,
} from "ts-morph";
import { getNodeByKind } from "../helper";

export const convertProps = (node: CallExpression, lang: string = "js") => {
  const expression = getNodeByKind(node, SyntaxKind.ObjectLiteralExpression);

  if (!expression) {
    throw new Error("props is not found.");
  }
  if (!Node.isObjectLiteralExpression(expression)) {
    throw new Error("props is not found.");
  }

  const properties = expression
    .getProperties()
    .filter((x) => x.getKind() === SyntaxKind.PropertyAssignment);

  const propsNode = getPropsNode(properties);

  return lang === "ts"
    ? convertToDefinePropsForTs(propsNode as PropertyAssignment)
    : convertToDefineProps(propsNode as PropertyAssignment);
};

const getPropsNode = (nodes: ObjectLiteralElementLike[]) => {
  const propsNode = nodes.find((x) => {
    const identifiler = (x as PropertyAssignment).getName();
    return identifiler === "props";
  });

  if (!propsNode) {
    throw new Error("props is not found.");
  }

  return propsNode;
};

const convertToDefineProps = (node: PropertyAssignment) => {
  const child = node.getInitializer();

  if (!child) {
    throw new Error("props is empty.");
  }

  if (Node.isObjectLiteralExpression(child)) {
    const properties = child.getProperties();
    const value = properties.map((x) => x.getText()).join(",");
    return `const props = defineProps({${value}})`;
  }

  return `const props = defineProps(${child.getFullText()})`;
};

// 以下 type-based declaration用

type PropType =
  | {
      type: "array";
      propertyName: string;
    }
  | {
      type: "typeOnly";
      propertyName: string;
      typeValue: string;
    }
  | {
      type: "object";
      propertyName: string;
      typeValue?: string;
      required?: boolean;
      defaultValue?: string | boolean;
    };

const convertToDefinePropsForTs = (node: PropertyAssignment) => {
  const child = node.getInitializer();

  if (Node.isObjectLiteralExpression(child)) {
    const properties = child.getProperties();

    const arr: PropType[] = properties.map((x) => {
      if (!Node.isPropertyAssignment(x)) {
        throw new Error("property not found.");
      }

      const propObj = x.getInitializer();
      if (Node.isObjectLiteralExpression(propObj)) {
        return {
          ...convertPropsWithObject(propObj),
          propertyName: x.getName(),
        };
      }
      return {
        type: "typeOnly",
        typeValue: propObj?.getText() ?? "",
        propertyName: x.getName(),
      };
    });

    const props = convertToTypeDefineProps(arr);

    return props;
  }

  if (!Node.isArrayLiteralExpression(child)) {
    throw new Error("props not found.");
  }
  return `const defineProps(${child.getText()});`;
};

const convertToTypeDefineProps = (props: PropType[]) => {
  const members = props
    .map((x) => {
      if (x.type === "array") {
        return;
      }
      if (x.type === "object") {
        return `${x.propertyName}${!x.required ? "?" : ""}: ${x.typeValue};`;
      }
      return `${x.propertyName}?: ${x.typeValue};`;
    })
    .filter(Boolean);

  const propType = `type Props = {${members.join("\n")}};`;

  const defineProps = `const props = defineProps<Props>();`;

  const hasDefault = props.find((x) => x.type === "object" && x.defaultValue);

  const defaultValueParams = props
    .map((x) => {
      if (x.type === "object" && x.defaultValue) {
        return `${x.propertyName}: ${x.defaultValue}`;
      }
    })
    .filter(Boolean)
    .join(",");

  const withDefaults = `const props = withDefaults(defineProps<Props>(), { ${defaultValueParams} });`;

  return hasDefault ? propType + withDefaults : propType + defineProps;
};

const convertPropsWithObject = (
  node: ObjectLiteralExpression
): {
  type: "object";
  typeValue: string;
  required?: boolean;
  defaultValue?: string | boolean;
} => {
  const properties = node.getProperties();

  const typeValue = getTypeValue(properties);

  const required = getPropsOption("required", properties);

  const defaultValue = getPropsOption("default", properties);

  return {
    type: "object",
    typeValue,
    required: required ? Boolean(required) : undefined,
    defaultValue,
  };
};

const getTypeValue = (properties: ObjectLiteralElementLike[]) => {
  const property = properties.find((x) => {
    if (!Node.isPropertyAssignment(x)) {
      throw new Error("props property not found.");
    }
    return x.getName() === "type";
  });

  if (!property) {
    throw new Error("props property not found.");
  }

  if (!Node.isPropertyAssignment(property)) {
    throw new Error("props property not found.");
  }

  const initializer = property.getInitializer();

  if (!initializer) {
    throw new Error("props property not found.");
  }

  return typeMapping[initializer.getText()];
};

const getPropsOption = (
  type: "required" | "default",
  properties: ObjectLiteralElementLike[]
) => {
  const property = properties.find((x) => {
    if (!Node.isPropertyAssignment(x)) {
      throw new Error("props property not found.");
    }
    return x.getName() === type;
  });

  if (!property) {
    return;
  }

  if (!Node.isPropertyAssignment(property)) {
    throw new Error("props property not found.");
  }

  const initializer = property.getInitializer();

  if (Node.isIdentifier(property.getInitializer())) {
    if (!initializer) {
      throw new Error("props property not found.");
    }
    return initializer.getText();
  } else {
    if (!initializer) {
      throw new Error("props property not found.");
    }
    if (
      isFalseKeyword(initializer.getKind()) ||
      isTrueKeyword(initializer.getKind())
    ) {
      return isTrueKeyword(initializer.getKind());
    }
    if (Node.isLiteralLike(initializer)) {
      return initializer.getText();
    }

    return initializer.getText();
  }
};

const isTrueKeyword = (kind: SyntaxKind) => {
  return kind === SyntaxKind.TrueKeyword;
};

const isFalseKeyword = (kind: number) => {
  return kind === SyntaxKind.FalseKeyword;
};

const typeMapping: Record<string, string> = {
  String: "string",
  Number: "number",
  Boolean: "boolean",
  Object: "any",
};
