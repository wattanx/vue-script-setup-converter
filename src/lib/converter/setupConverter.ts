import { getNodeByKind } from "./../helper";
import { CallExpression, SyntaxKind } from "ts-morph";

export const convertSetup = (node: CallExpression) => {
  const setupNode = getNodeByKind(node, SyntaxKind.MethodDeclaration);

  if (!setupNode) {
    throw new Error("setup is not found.");
  }

  const blockNode = getNodeByKind(setupNode, SyntaxKind.Block);

  if (!blockNode) {
    throw new Error("setup is not found.");
  }

  return blockNode
    .forEachChildAsArray()
    .filter((x) => x.getKind() !== SyntaxKind.ReturnStatement)
    .map((x) => x.getFullText())
    .join("");
};
