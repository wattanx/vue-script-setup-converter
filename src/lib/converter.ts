import {
  CallExpression,
  ScriptTarget,
  SyntaxKind,
  Project,
  Node,
  ImportDeclaration,
} from "ts-morph";
import { parse } from "@vue/compiler-sfc";
import prettier from "prettier";
import parserTypeScript from "prettier/parser-typescript";
import { getNodeByKind } from "./helper";
import { convertProps } from "./converter/propsConverter";
import { convertSetup } from "./converter/setupConverter";
import { convertEmits } from "./converter/emitsConverter";

export const convertSrc = (input: string) => {
  const {
    descriptor: { script },
  } = parse(input);

  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.Latest,
    },
    useInMemoryFileSystem: true,
  });

  const sourceFile = project.createSourceFile("s.tsx", script?.content ?? "");
  const lang = script?.lang ?? "js";

  const importDeclaration = getNodeByKind(
    sourceFile,
    SyntaxKind.ImportDeclaration
  ) as ImportDeclaration;

  const importStatement = importDeclaration.getText() ?? "";

  const callexpression = getNodeByKind(sourceFile, SyntaxKind.CallExpression);

  if (!callexpression) {
    throw new Error("defineComponent is not found.");
  }
  if (!Node.isCallExpression(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  if (!isDefineComponent(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  const props = convertProps(callexpression, lang) ?? "";
  const emits = convertEmits(callexpression, lang) ?? "";
  const statement = convertSetup(callexpression) ?? "";

  const formatedText = prettier.format(
    importStatement + props + emits + statement,
    {
      parser: "typescript",
      plugins: [parserTypeScript],
    }
  );

  return formatedText;
};

const isDefineComponent = (node: CallExpression) => {
  if (!Node.isIdentifier(node.getExpression())) {
    return false;
  }

  return node.getExpression().getText() === "defineComponent";
};
