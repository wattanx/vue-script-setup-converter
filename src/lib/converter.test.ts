import { expect, describe, it } from "vitest";
import { convertSrc } from "./converter";
import srcJs from "../assets/template/composition-api.txt?raw";
import srcTs from "src/assets/template/composition-api-ts.txt?raw";

describe("lang=js", () => {
  it("convert", () => {
    const output = convertSrc(srcJs);
    expect(output).toMatchSnapshot();
  });
});

describe("lang=ts", () => {
  it("type-base declaration", () => {
    const output = convertSrc(srcTs);
    expect(output).toMatchSnapshot();
  });
});
