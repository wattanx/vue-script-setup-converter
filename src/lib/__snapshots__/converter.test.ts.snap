// Vitest Snapshot v1

exports[`lang=js > convert 1`] = `
"import { defineComponent, toRefs, computed, ref } from \\"vue\\";
const props = defineProps({
  msg: {
    type: String,
    required: true,
    default: \\"HelloWorld\\",
  },
  foo: {
    type: String,
  },
});
const emit = defineEmits({ change: null });
const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + \\"- HelloWorld\\");

const count = ref(0);
"
`;

exports[`lang=ts > type-base declaration 1`] = `
"import { defineComponent, toRefs, computed, ref } from \\"vue\\";
type Props = { msg: string; foo?: string };
const props = withDefaults(defineProps<Props>(), { msg: \\"HelloWorld\\" });
const emit = defineEmits<{ (e: \\"change\\", value: number): void }>();
const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + \\"- HelloWorld\\");

const count = ref(0);
"
`;
